File: Application/src/main/java/com/shareyourproxy/api/rx/command/DeleteUserChannelCommand.java
package com.shareyourproxy.api.rx.command;

import android.app.Service;
import android.os.Parcel;
import android.support.annotation.NonNull;

import com.shareyourproxy.api.domain.model.Channel;
import com.shareyourproxy.api.domain.model.User;
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback;

import java.util.List;

import static com.shareyourproxy.api.rx.RxUserChannelSync.deleteChannel;

/**
 * Delete a channel associated with a user.
 */
public class DeleteUserChannelCommand extends BaseCommand {

    public static final Creator<DeleteUserChannelCommand> CREATOR =
        new Creator<DeleteUserChannelCommand>() {
            @Override
            public DeleteUserChannelCommand createFromParcel(Parcel in) {
                return new DeleteUserChannelCommand(in);
            }

            @Override
            public DeleteUserChannelCommand[] newArray(int size) {
                return new DeleteUserChannelCommand[size];
            }
        };
    private final static java.lang.ClassLoader CL = DeleteUserChannelCommand.class.getClassLoader();
    public final Channel channel;
    public final int position;
    public final User user;

    public DeleteUserChannelCommand(@NonNull User user, @NonNull Channel channel, int position) {
        this.user = user;
        this.channel = channel;
        this.position = position;
    }

    private DeleteUserChannelCommand(Parcel in) {
        this((User) in.readValue(CL), (Channel) in.readValue(CL), (int) in.readValue(CL));
    }

    @Override
    public List<EventCallback> execute(Service service) {
        return deleteChannel(service, user, channel, position);
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeValue(user);
        dest.writeValue(channel);
        dest.writeValue(position);
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/api/rx/command/DeleteUserChannelCommand.kt
package com.shareyourproxy.api.rx.command

import android.app.Service
import android.os.Parcel
import android.os.Parcelable
import com.shareyourproxy.api.domain.model.Channel
import com.shareyourproxy.api.domain.model.User
import com.shareyourproxy.api.rx.RxUserChannelSync.deleteChannel
import com.shareyourproxy.api.rx.command.eventcallback.EventCallback


/**
 * Delete a channel associated with a user.
 */
internal final class DeleteUserChannelCommand(val user: User, val channel: Channel, val position: Int) : BaseCommand() {

    private constructor(parcel: Parcel) : this(parcel.readValue(CL) as User, parcel.readValue(CL) as Channel, parcel.readValue(CL) as Int)

    override fun execute(service: Service): EventCallback {
        return deleteChannel(service, user, channel, position)
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeValue(user)
        dest.writeValue(channel)
        dest.writeValue(position)
    }

    companion object {
        private val CL = DeleteUserChannelCommand::class.java.classLoader
        val CREATOR: Parcelable.Creator<DeleteUserChannelCommand> = object : Parcelable.Creator<DeleteUserChannelCommand> {
            override fun createFromParcel(parcel: Parcel): DeleteUserChannelCommand {
                return DeleteUserChannelCommand(parcel)
            }

            override fun newArray(size: Int): Array<DeleteUserChannelCommand?> {
                return arrayOfNulls(size)
            }
        }
    }
}
